plugin/pkg/utils/namespacedcloudprofile.go

```diff
@@ -0,0 +1,87 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package utils
+
+import (
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/gardener/gardener/pkg/apis/core"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	gardencorev1beta1listers "github.com/gardener/gardener/pkg/client/core/listers/core/v1beta1"
+	"github.com/gardener/gardener/pkg/utils/gardener"
+)
+
+// GetCloudProfile determines whether a given CloudProfile (plus namespace) is a valid NamespacedCloudProfile or a CloudProfile and returns the appropriate object
+func GetCloudProfile(cloudProfileLister gardencorev1beta1listers.CloudProfileLister, NamespacedCloudProfileLister gardencorev1beta1listers.NamespacedCloudProfileLister, cp *core.CloudProfileReference, name *string, namespace string) (*gardencorev1beta1.CloudProfile, error) {
+	cloudProfileReference := gardener.BuildCloudProfileReference(name, cp)
+
+	if cloudProfileReference.Kind == constants.CloudProfileReferenceKindNamespacedCloudProfile {
+		var ncpErr error
+		namespacedCloudProfile, ncpErr := NamespacedCloudProfileLister.NamespacedCloudProfiles(namespace).Get(cloudProfileReference.Name)
+		if ncpErr == nil {
+			return &gardencorev1beta1.CloudProfile{Spec: namespacedCloudProfile.Status.CloudProfileSpec}, nil
+		}
+		if !apierrors.IsNotFound(ncpErr) {
+			return nil, ncpErr
+		}
+	}
+
+	cloudProfile, err := cloudProfileLister.Get(cloudProfileReference.Name)
+	if err == nil {
+		return cloudProfile, nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	return nil, fmt.Errorf("could not get cloud profile: %+v", err.Error())
+}
+
+// ValidateCloudProfileChanges validates that CloudProfiles referenced are all the same CloudProfile root node or children thereof (NamespacedCloudProfiles with the root node as parent)
+func ValidateCloudProfileChanges(cloudProfileLister gardencorev1beta1listers.CloudProfileLister, namespacedCloudProfileLister gardencorev1beta1listers.NamespacedCloudProfileLister, newShootSpec, oldShootSpec core.ShootSpec, namespace string) error {
```

Comment:
```suggestion
// ValidateImmutableCloudProfile validates the referenced CloudProfile in new and old shoot is the same.
func ValidateCloudProfileChanges(cloudProfileLister gardencorev1beta1listers.CloudProfileLister, namespacedCloudProfileLister gardencorev1beta1listers.NamespacedCloudProfileLister, newShoot, oldShoot *core.Shoot error {
```

Reply:
as discussed, immutability is not the term we are looking for here. Will change the functionality to restrict changes to only CloudProfile → descendant NamespacedCloudProfile.